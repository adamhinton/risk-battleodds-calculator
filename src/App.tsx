// TODO: Refactor, particularly make the simulation more DRY and figure out how to best replicate userInputs in runSingleSimulation

function App() {
  return (
    <div className="App">
      <h1>Adam Hinton</h1>
    </div>
  );
}

function generatePlayerRolls (playerType: PlayerType, diceRolls: 1 | 2 | number): PlayerRolls {
  const generateSortedRolls = (diceRolls: 1 | 2 | 3): PlayerRolls =>{
    const solution = [] 
    for(let i=0; i<diceRolls; i++){
      solution.push(randomIntFromInterval(1, 6))
    }
    // TODO: Refactor this, I shouldn't need this type assertion
   return sortPlayerRolls(solution)
  }

  // if one unit, doesn't matter if attacker or defender, they get one dice roll
  if(diceRolls === 1){
    return generateSortedRolls(1)
  }

  // They're a defender and we already know they don't have one unit, so they get two dice rolls
  if(playerType === 'defender'){
    return generateSortedRolls(2)
  }

  // We know they're an attacker now. So if they have two units they get two rolls, otherwise they get three rolls
  return diceRolls === 2? generateSortedRolls(2) : generateSortedRolls(3)
}

export default App;

// TYPE DEFINITIONS
type PlayerCounts = {
  attackerCount: number;
  defenderCount: number;
}

type Results = {
  attackerOccupies: number;
  defenderHolds: number;
  // averageUnitsLeft: {
  //   attackers: number;
  //   defenders: number;
  // }
}

type UserInputs = {
  playerCounts: PlayerCounts;
  numSimulations: number;
}

type PlayerType = 'attacker' | 'defender'

type AttackerRolls = [number, number? , number?]
type DefenderRolls = [number, number?]
type PlayerRolls = AttackerRolls | DefenderRolls

// USEFUL VARIABLES
const userInputs: UserInputs = {
  playerCounts: {
   attackerCount: 20,
   defenderCount: 20
  },
  numSimulations: 1
} 

const results: Results = {
  attackerOccupies: 0,
  defenderHolds: 0
}

// UTILS
const randomIntFromInterval = (min: Readonly<number>, max: Readonly<number>) =>{ // min and max included 
  return Math.floor(Math.random() * (max - min + 1) + min)
}

const sortPlayerRolls = (rolls: PlayerRolls) =>{
  return rolls.sort((a , b ) =>{ return b!  - a! })
}

console.time('a')
for(let i=0; i<userInputs.numSimulations; i++){
  runSingleSimulation( {...userInputs.playerCounts } )
}

console.log('results:', results)
console.timeEnd('a')
console.log('userInputs:', userInputs)

function runSingleSimulation (playerCounts: PlayerCounts): void {

// One run through of simulation
while ( playerCounts.attackerCount > 0 && playerCounts.defenderCount > 0 ){

const attackerRolls = generatePlayerRolls('attacker', playerCounts.attackerCount)
const defenderRolls = generatePlayerRolls('defender', playerCounts.defenderCount)

console.log('attackerRolls, defenderRolls:', attackerRolls, defenderRolls)

// const attackerRolls = sortPlayerRolls(attackerRolls)
// const defenderRollsSorted = sortPlayerRolls(defenderRolls)

// if diceRolls is 1 or 2 then follow certain logic. If more than 2 (aka its type is number) then follow other logic.


// Now for the attack 
if(attackerRolls[0] > defenderRolls[0]){
  playerCounts.defenderCount --
}
else {
  playerCounts.attackerCount --
}

if(playerCounts.attackerCount > 1 && playerCounts.defenderCount > 1 ){
  if(attackerRolls[1]! > defenderRolls[1]!){
  playerCounts.defenderCount --
}
  else {
  playerCounts.attackerCount --
   }
}

// results
if(playerCounts.attackerCount === 0){
  results.defenderHolds ++
}
else if (playerCounts.defenderCount === 0){
  results.attackerOccupies ++
}
}
}