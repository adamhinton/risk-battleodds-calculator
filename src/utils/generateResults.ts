export type UserInputs = {
	attackerCount: number;
	defenderCount: number[];
	numSimulations: number;
	stopAt: number;
};

type PlayerType = "attacker" | "defender";

export type AttackerRolls = [number, number?, number?];
type DefenderRolls = [number, number?];
type PlayerRolls = AttackerRolls | DefenderRolls;

export type Results = {
	attackerOccupies: number;
	defenderHolds: number;
	averageAttackersLeft?: number | null;
	averageDefendersLeft?: number | null;
};

export type PlayerCounts = {
	attackerCount: number;
	defenderCount: number[];
};

type SingleSimResults = {
	attackersLeft: number;
	defendersLeft: number;
	attackerOccupies: boolean;
	defenderHolds: boolean;
};

const generateResults = (
	userInputs: Readonly<UserInputs>
): Readonly<Results> => {
	const results: Results = {
		attackerOccupies: 0,
		defenderHolds: 0,
		averageAttackersLeft: null,
		averageDefendersLeft: null,
	};

	let totalAttackersLeft = 0;
	let totalDefendersLeft = 0;

	for (let i = 0; i < userInputs.numSimulations; i++) {
		const singleSimResults: SingleSimResults = runSingleSimulation(
			{
				attackerCount: userInputs.attackerCount,
				defenderCount: [...userInputs.defenderCount],
			},
			userInputs.stopAt
		);

		if (singleSimResults.attackerOccupies === true) {
			results.attackerOccupies++;
			totalAttackersLeft += singleSimResults.attackersLeft;
		} else if (singleSimResults.defenderHolds === true) {
			results.defenderHolds++;
			totalDefendersLeft += singleSimResults.defendersLeft;
		}
	}
	results.averageAttackersLeft = totalAttackersLeft / userInputs.numSimulations;
	results.averageDefendersLeft = totalDefendersLeft / userInputs.numSimulations;
	return results;
};

// UTILS

export function runSingleSimulation(
	playerCounts: PlayerCounts,
	stopAt: number
): Readonly<SingleSimResults> {
	let singleSimResults: SingleSimResults = {
		attackersLeft: 0,
		defendersLeft: 0,
		attackerOccupies: false,
		defenderHolds: false,
	};

	// loop through playerCounts.defenderCount

	playerCounts.defenderCount.forEach((defender: number, index) => {
		while (playerCounts.defenderCount[index] > 0) {
			// If attacker has reached the number of troops where they have specified to stop:
			// Defender holds, and the number of attackers and defenders left are added to their averages.
			if (playerCounts.attackerCount <= stopAt) {
				singleSimResults.defenderHolds = true;
				singleSimResults.attackersLeft = playerCounts.attackerCount;
				const totalOfDefenders = playerCounts.defenderCount.reduce(
					(partialSum, a) => partialSum + a,
					0
				);
				singleSimResults.defendersLeft = totalOfDefenders;
				break;
			}

			const attackerRolls = generatePlayerRolls(
				"attacker",
				playerCounts.attackerCount
			);
			const defenderRolls = generatePlayerRolls(
				"defender",
				playerCounts.defenderCount[index]
			);

			// Now for the attack
			// Attacker wins first dice roll
			if (attackerRolls[0] > defenderRolls[0]) {
				playerCounts.defenderCount[index]--;
			}
			// Attacker loses first dice roll
			else {
				playerCounts.attackerCount--;
			}

			// If there is a second dice roll
			if (
				attackerRolls[1] &&
				// Found the problem. We just potentially did defendercount[index]-- so this may not be accurate. So we were overestimating defenders by doing this.
				defenderRolls[1]
			) {
				// Attacker wins second dice roll
				if (attackerRolls[1]! > defenderRolls[1]!) {
					playerCounts.defenderCount[index]--;
				}
				// Attacker loses second dice roll
				else {
					playerCounts.attackerCount--;
				}
			}

			// results
			//defender wins entire battle
			if (playerCounts.attackerCount < 1) {
				singleSimResults.defenderHolds = true;
				const totalOfDefenders = playerCounts.defenderCount.reduce(
					(partialSum, a) => partialSum + a,
					0
				);
				singleSimResults.defendersLeft = totalOfDefenders;
				break;
			}
			// defender occupies current territory but there are more defending territories left
			if (
				playerCounts.defenderCount[index] < 1 &&
				index < playerCounts.defenderCount.length - 1
			) {
				// attacker loses a troop because they have to occupy the conquered territory with one troop
				playerCounts.attackerCount--;
			}
			// Attacker occupies last defending territory and wins the whole battle
			if (
				playerCounts.defenderCount[index] === 0 &&
				index === playerCounts.defenderCount.length - 1
			) {
				singleSimResults.attackerOccupies = true;
				singleSimResults.attackersLeft = playerCounts.attackerCount;
			}
		}
	});
	return singleSimResults;
}

function randomIntFromInterval(
	min: Readonly<number>,
	max: Readonly<number>
): number {
	// min and max included
	return Math.floor(Math.random() * (max - min + 1) + min);
}

export function sortPlayerRolls(rolls: PlayerRolls): PlayerRolls {
	return rolls.sort((a, b) => {
		return b! - a!;
	});
}

export function generatePlayerRolls(
	playerType: Readonly<PlayerType>,
	diceRolls: Readonly<1 | 2 | number>
): PlayerRolls {
	const generateSortedRolls = (diceRolls: Readonly<1 | 2 | 3>): PlayerRolls => {
		const solution = [];
		for (let i = 0; i < diceRolls; i++) {
			solution.push(randomIntFromInterval(1, 6));
		}
		// TODO: Refactor this, I shouldn't need this type assertion
		return sortPlayerRolls(solution as PlayerRolls);
	};

	// if one unit, doesn't matter if attacker or defender, they get one dice roll
	if (diceRolls === 1) {
		return generateSortedRolls(1);
	}

	// They're a defender and we already know they don't have one unit, so they get two dice rolls
	if (playerType === "defender") {
		return generateSortedRolls(2);
	}

	// We know they're an attacker now. So if they have two units they get two rolls, otherwise they get three rolls
	return diceRolls === 2 ? generateSortedRolls(2) : generateSortedRolls(3);
}

export default generateResults;
