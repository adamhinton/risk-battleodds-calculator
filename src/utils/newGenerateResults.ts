export type UserInputs = {
  attackerCount: number;
  defenderCount: number[];
  numSimulations: number;
};

type PlayerType = "attacker" | "defender";

type AttackerRolls = [number, number?, number?];
type DefenderRolls = [number, number?];
type PlayerRolls = AttackerRolls | DefenderRolls;

export type Results = {
  attackerOccupies: number;
  defenderHolds: number;
  averageAttackersLeft?: number | null;
  averageDefendersLeft?: number | null;
};

const newGenerateResults = (userInputs: UserInputs): Results => {
  const results: Results = {
    attackerOccupies: 0,
    defenderHolds: 0,
    averageAttackersLeft: null,
    averageDefendersLeft: null,
  };

  let totalAttackersLeft = 0;
  let totalDefendersLeft = 0;

  return results;
};

export default newGenerateResults;
