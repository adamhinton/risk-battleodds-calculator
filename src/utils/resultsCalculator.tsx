// TYPE DEFINITIONS
type PlayerCounts = {
  attackerCount: number;
  defenderCount: number;
};

export type Results = {
  attackerOccupies: number;
  defenderHolds: number;
  averageAttackersLeft?: number;
  averageDefendersLeft?: number;
};

export type UserInputs = {
  attackerCount: number;
  defenderCount: number;
  numSimulations: number;
};

console.log("don't forget to make all items in Results required");

type PlayerType = "attacker" | "defender";

type AttackerRolls = [number, number?, number?];
type DefenderRolls = [number, number?];
type PlayerRolls = AttackerRolls | DefenderRolls;

const generateResults = (userInputs: UserInputs) => {
  // USEFUL VARIABLES
  const results: Results = {
    attackerOccupies: 0,
    defenderHolds: 0,
  };

  let totalAttackersLeft = 0;
  let totalDefendersLeft = 0;

  // UTILS

  console.time("a");
  for (let i = 0; i < userInputs.numSimulations; i++) {
    runSingleSimulation({
      attackerCount: userInputs.attackerCount,
      defenderCount: userInputs.defenderCount,
    });
  }

  console.log("results:", results);
  console.timeEnd("a");
  console.log("userInputs:", userInputs);

  function runSingleSimulation(playerCounts: PlayerCounts): void {
    // One run through of simulation
    while (playerCounts.attackerCount > 0 && playerCounts.defenderCount > 0) {
      const attackerRolls = generatePlayerRolls(
        "attacker",
        playerCounts.attackerCount
      );
      const defenderRolls = generatePlayerRolls(
        "defender",
        playerCounts.defenderCount
      );

      // Now for the attack
      if (attackerRolls[0] > defenderRolls[0]) {
        playerCounts.defenderCount--;
      } else {
        playerCounts.attackerCount--;
      }

      if (playerCounts.attackerCount > 1 && playerCounts.defenderCount > 1) {
        if (attackerRolls[1]! > defenderRolls[1]!) {
          playerCounts.defenderCount--;
        } else {
          playerCounts.attackerCount--;
        }
      }

      // results
      if (playerCounts.attackerCount === 0) {
        results.defenderHolds++;
        totalDefendersLeft += playerCounts.defenderCount;
      } else if (playerCounts.defenderCount === 0) {
        results.attackerOccupies++;
        totalAttackersLeft += playerCounts.attackerCount;
      }
    }
    // console.log("totalAttackersLeft:", totalAttackersLeft);
    // console.log("totalDefendersLeft:", totalDefendersLeft);
  }

  return results;
};

function sortPlayerRolls(rolls: PlayerRolls): PlayerRolls {
  return rolls.sort((a, b) => {
    return b! - a!;
  });
}

function randomIntFromInterval(min: Readonly<number>, max: Readonly<number>) {
  // min and max included
  return Math.floor(Math.random() * (max - min + 1) + min);
}

function generatePlayerRolls(
  playerType: PlayerType,
  diceRolls: 1 | 2 | number
): PlayerRolls {
  const generateSortedRolls = (diceRolls: 1 | 2 | 3): PlayerRolls => {
    const solution = [];
    for (let i = 0; i < diceRolls; i++) {
      solution.push(randomIntFromInterval(1, 6));
    }
    // TODO: Refactor this, I shouldn't need this type assertion
    return sortPlayerRolls(solution as PlayerRolls);
  };

  // if one unit, doesn't matter if attacker or defender, they get one dice roll
  if (diceRolls === 1) {
    return generateSortedRolls(1);
  }

  // They're a defender and we already know they don't have one unit, so they get two dice rolls
  if (playerType === "defender") {
    return generateSortedRolls(2);
  }

  // We know they're an attacker now. So if they have two units they get two rolls, otherwise they get three rolls
  return diceRolls === 2 ? generateSortedRolls(2) : generateSortedRolls(3);
}

export default generateResults;
