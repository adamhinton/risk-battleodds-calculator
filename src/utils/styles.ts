export type Colors = {
	primary: string;
	accent: {
		primary: string;
		secondary: string;
	};
	dark: {
		background: string;
		header: string;
		form: string;
		text: string;
		input: string;
		headerText: string;
	};
	light: {
		background: string;
		header: string;
		form: string;
		text: string;
		input: string;
		headerText: string;
	};
};

export const colors: Colors = {
	primary: "#1976d2",
	accent: {
		primary: "#64b5f6",
		secondary: "#4dd0e1",
	},
	dark: {
		background: "#1e272e",
		header: "#263238",
		form: "#2d3c42",
		text: "#ffffff",
		input: "#ffffff",
		headerText: "#ffffff",
	},
	light: {
		background: "#f0f0f0",
		header: "#e0e0e0",
		form: "#ffffff",
		text: "#333333",
		input: "#333333",
		headerText: "#333333",
	},
};

export const typography = {
	fontFamily: "'Roboto', sans-serif",
	h1: {
		fontSize: "2rem",
		fontWeight: 700,
	},
	h2: {
		fontSize: "1.5rem",
		fontWeight: 600,
	},
	h4: {
		fontSize: "1.2rem",
		fontWeight: 600,
	},
	h6: {
		fontSize: "1rem",
		fontWeight: 600,
	},
	body: {
		fontSize: "0.9rem",
		fontWeight: 400,
	},
	small: {
		fontSize: "0.8rem",
		fontWeight: 400,
	},
};

export const spacing = {
	paddingSmall: "8px",
	paddingMedium: "16px",
	paddingLarge: "24px",
	marginSmall: "8px",
	marginMedium: "16px",
	marginLarge: "24px",
	borderRadius: "8px",
};
