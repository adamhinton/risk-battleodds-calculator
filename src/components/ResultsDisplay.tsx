import { Results } from "../utils/generateResults";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";
import { Typography } from "@mui/material";
import styled from "styled-components";
import EmojiEventsIcon from "@mui/icons-material/EmojiEvents";

type ResultsDisplayProps = {
	results: Results;
};

const ResultsDisplay = (props: ResultsDisplayProps) => {
	const { results } = props;
	const { attackerOccupies, defenderHolds } = results;

	const totalSimulations = attackerOccupies + defenderHolds;
	const attackerWinPercent = (attackerOccupies / totalSimulations) * 100;
	const defenderWinPercent = (defenderHolds / totalSimulations) * 100;

	return (
		<StyledCard data-testid="results-display-section">
			<CardContent>
				<StyledHeader>
					<h2>Results</h2>
					<StyledEmojiIcon />
				</StyledHeader>
				<StyledResult>
					<Typography component="p">
						Attacker Occupies:{" "}
						<StyledPercent>{attackerWinPercent.toFixed(1)}%</StyledPercent>
					</Typography>
					<Typography component="p">
						Average Attackers Left:{" "}
						<StyledValue>
							{results.averageAttackersLeft?.toFixed(1)}
						</StyledValue>
					</Typography>
				</StyledResult>
				<StyledResult>
					<Typography component="p">
						Defender Holds:{" "}
						<StyledPercent>{defenderWinPercent.toFixed(1)}%</StyledPercent>
					</Typography>
					<Typography component="p">
						Average Defenders Left:{" "}
						<StyledValue>
							{results.averageDefendersLeft?.toFixed(1)}
						</StyledValue>
					</Typography>
				</StyledResult>
			</CardContent>
		</StyledCard>
	);
};

export default ResultsDisplay;

const StyledCard = styled(Card)`
	&& {
		margin-top: 20px;
		padding: 20px 30px;
		background-color: ${({ theme }) => theme.customTheming.formAndInputsBGC};
		color: ${({ theme }) => theme.customTheming.formTextColor};
	}
`;

const StyledHeader = styled.div`
	display: flex;
	align-items: center;
	justify-content: space-between;
	margin-bottom: 15px;
`;

const StyledEmojiIcon = styled(EmojiEventsIcon)`
	color: ${({ theme }) => theme.customTheming.formTextColor};
	font-size: 2rem;
`;

const StyledResult = styled.div`
	margin-top: 15px;
	text-align: center;

	p {
		margin-bottom: 10px;
	}
`;

const StyledPercent = styled.span`
	font-size: 1.1rem;
	color: ${({ theme }) => theme.customTheming.headerTextColor};
`;

const StyledValue = styled.span`
	font-size: 1.1rem;
`;
