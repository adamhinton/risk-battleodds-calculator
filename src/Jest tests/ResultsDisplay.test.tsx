import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import ResultsDisplay from "../components/ResultsDisplay";
import { Results } from "../utils/resultsCalculator";
import { darkTheme } from "../App";
import { ThemeProvider } from "styled-components";

const fakeResults: Results = {
	attackerOccupies: 1234,
	defenderHolds: 1343,
	averageAttackersLeft: 18.9545,
	averageDefendersLeft: 28.5454,
};

const secondFakeResults: Results = {
	attackerOccupies: 100,
	defenderHolds: 100,
	averageAttackersLeft: 16,
	averageDefendersLeft: 8,
};

test("[1] Renders without errors", () => {
	render(
		<ThemeProvider theme={darkTheme}>
			<ResultsDisplay results={fakeResults} />
		</ThemeProvider>
	);
});

test("[2] Displays text of results as expected", () => {
	render(
		<ThemeProvider theme={darkTheme}>
			<ResultsDisplay results={fakeResults} />
		</ThemeProvider>
	);

	const h2 = screen.getByTestId("results-h2");
	const attackerOccupies = screen.getByTestId("results-attacker-occupies");
	const avgAttackersLeft = screen.getByTestId("results-avg-attackers-left");
	const defenderHolds = screen.getByTestId("results-defenderholds");
	const avgDefendersLeft = screen.getByTestId("results-avg-defenders-left");

	expect(h2).toHaveTextContent("Results");
	expect(attackerOccupies).toHaveTextContent("Attacker Occupies: 47.9%");
	expect(avgAttackersLeft).toHaveTextContent("Average Attackers Left: 19.0");
	expect(defenderHolds).toHaveTextContent("Defender Holds: 52.1%");
	expect(avgDefendersLeft).toHaveTextContent("Average Defenders Left: 28.5");
});

test("[3] Updates when passed new props", () => {
	const { rerender } = render(
		<ThemeProvider theme={darkTheme}>
			<ResultsDisplay results={fakeResults} />
		</ThemeProvider>
	);

	rerender(
		<ThemeProvider theme={darkTheme}>
			<ResultsDisplay results={secondFakeResults} />
		</ThemeProvider>
	);

	const avgAttackersLeft = screen.getByTestId("results-avg-attackers-left");
	expect(avgAttackersLeft).toHaveTextContent("Average Attackers Left: 16.0");
	expect(avgAttackersLeft).not.toHaveTextContent(
		"Average Attackers Left: 19.0"
	);
});

test("[4] Matches screenshot from 3.1.2023", () => {
	const tree = renderer
		.create(
			<ThemeProvider theme={darkTheme}>
				<ResultsDisplay results={secondFakeResults} />
			</ThemeProvider>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
