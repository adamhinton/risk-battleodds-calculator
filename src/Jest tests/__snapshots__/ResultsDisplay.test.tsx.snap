// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[4] Matches screenshot from 3.1.2023 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root sc-beySbM lajeib css-bhp9pd-MuiPaper-root-MuiCard-root"
  data-testid="results-display-section"
>
  <div
    className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
  >
    <h2
      data-testid="results-h2"
    >
      Results
    </h2>
    <div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        data-testid="results-attacker-occupies"
      >
        Attacker Occupies: 
        50.0
        %
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        data-testid="results-avg-attackers-left"
      >
        Average Attackers Left: 
        16.0
      </p>
    </div>
    <div>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        data-testid="results-defenderholds"
      >
        Defender Holds: 
        50.0
        %
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        data-testid="results-avg-defenders-left"
      >
        Average Defenders Left: 
        8.0
      </p>
    </div>
  </div>
</div>
`;
